File: packages/sandbox-runtime/src/sandbox_runtime/runtime_manifest.py
"""Validated sandbox runtime compatibility manifest."""

import json
import re
from pathlib import Path
from typing import TypedDict, cast


class RuntimeManifest(TypedDict):
    runtimeVersion: str
    generation: int
    minimumCompatibleGeneration: int
    minimumPreservationGeneration: int
    minimumRebuildGeneration: int
    # Per-harness floors for prebuilt-image selection: a harness whose runtime
    # support arrived later than the global floor names its own generation.
    harnessMinimumGeneration: dict[str, int]


_MANIFEST_PATH = Path(__file__).with_name("runtime_manifest.json")
RUNTIME_MANIFEST = cast("RuntimeManifest", json.loads(_MANIFEST_PATH.read_text()))
RUNTIME_VERSION = RUNTIME_MANIFEST["runtimeVersion"]
_VERSION_MATCH = re.match(r"^v(\d+)", RUNTIME_VERSION)

if not _VERSION_MATCH or int(_VERSION_MATCH.group(1)) != RUNTIME_MANIFEST["generation"]:
    raise RuntimeError("Sandbox runtime manifest version and generation disagree")

File: packages/control-plane/src/sandbox/runtime-manifest.ts
import runtimeManifest from "../../../sandbox-runtime/src/sandbox_runtime/runtime_manifest.json";

const parsedGeneration = /^v(\d+)/.exec(runtimeManifest.runtimeVersion)?.[1];
if (Number(parsedGeneration) !== runtimeManifest.generation) {
  throw new Error("Sandbox runtime manifest version and generation disagree");
}

export const SANDBOX_RUNTIME_VERSION = runtimeManifest.runtimeVersion;
export const SANDBOX_RUNTIME_GENERATION = runtimeManifest.generation;
export const MIN_COMPATIBLE_RUNTIME_GENERATION = runtimeManifest.minimumCompatibleGeneration;
export const MIN_REBUILD_RUNTIME_GENERATION = runtimeManifest.minimumRebuildGeneration;
